### ISSUE TITLE
Cannot compile C++ CLR project

### ISSUE THREAD

**Issue by korhun** [tesseract-ocr/tesseract#2866]
I was able to successfully build a C++ Common Language RunTime Support (/clr) project and use it with C# with the older versions. However with the latest updates, I get this build error:

scrollview.h(406,8): error C2039: 'mutex': is not a member of 'std'

It seems that mutex cannot be used in a CLR project. So is there no way to build the Tesseract 5 in a CLR project?

**Comment by amitdo** [comment]
https://github.com/tesseract-ocr/tesseract/pull/2567

I don't think we should bend our code to make it fit to non-standard Microsoft extension.

@stweil, what's your opinion?

**Comment by stweil** [comment]
@korhun, https://docs.microsoft.com/de-de/cpp/standard-library/mutex-class-stl?view=vs-2019 clearly says that MS implements the standard with `std::mutex`, and it is made available by including the header file `mutex`.

Could it be that the correct include files are not found because you have other files with identical name in your include search path?

**Comment by amitdo** [comment]
He use the `/clr` flag which break the C++ standard.

https://docs.microsoft.com/en-us/cpp/standard-library/mutex?view=vs-2019

>In code that is compiled by using /clr, this header is blocked.

**Comment by stweil** [comment]
> So is there no way to build the Tesseract 5 in a CLR project?

@korhun, is it necessary to compile all files with `/clr`? If not, the easy solution would be to compile the Tesseract code without that flag.

`scrollview.h` only uses a pointer to a `std::mutex`, so you could try to replace the include statement for `mutex` by a forward declaration. If that works, you could send a pull request which updates the Tesseract code. But I am afraid that more code locations will follow with the same problem.

**Comment by stweil** [comment]
See also https://stackoverflow.com/questions/15821942/how-to-implement-a-unmanaged-thread-safe-collection-when-i-get-this-error-mute/15822678#15822678.

**Comment by korhun** [comment]
Hi, thank you very much for your interest to my question. You guys do a great job, I appreciate very much.

I'm not experienced in c and c++. After I've opened this issue, I've learned that we can export functions from a native c++ project using extern "C". So I was able to successfully build Tesseract50.dll with my custom extern functions, without /clr, using Visual Studio 2019, on both x64 and Win32. And could able use your library in a C# project without any problems.

So I guess this issue can be closed.
